plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs

```diff
@@ -139,86 +168,20 @@ with this, since you can't really specify a polymorphic type in a type applicati
-    - If we make any mistakes creating the Core expression
-    -}
-
-    -- Get the names of functions/types that we need for our expression
-    upio <- GHC.lookupId =<< thNameToGhcNameOrFail 'unsafePerformIO
-    bsTc <- GHC.lookupTyCon =<< thNameToGhcNameOrFail ''BS.ByteString
-    upal <- GHC.lookupId =<< thNameToGhcNameOrFail 'BSUnsafe.unsafePackAddressLen
-
-    -- We construct the following expression:
-    -- unsafePerformIO $ unsafePackAddressLen <length as int literal> <data as string literal address>
-    -- This technique gratefully borrowed from the file-embed package
-
-    -- The flags here are so GHC can check whether the int is in range for the current platform.
-    let lenLit = GHC.mkIntExpr flags $ fromIntegral $ BS.length bs
-    -- This will have type Addr#, which is right for unsafePackAddressLen
-    let bsLit = GHC.Lit (GHC.LitString bs)
-    let upaled = GHC.mkCoreApps (GHC.Var upal) [lenLit, bsLit]
-    let upioed = GHC.mkCoreApps (GHC.Var upio) [GHC.Type (GHC.mkTyConTy bsTc), upaled]
-
-    pure upioed
-
--- | Make a 'BuiltinNameInfo' mapping the given set of TH names to their
--- 'GHC.TyThing's for later reference.
-makePrimitiveNameInfo :: [TH.Name] -> GHC.CoreM BuiltinNameInfo
-makePrimitiveNameInfo names = do
-    infos <- forM names $ \name -> do
-        ghcName <- thNameToGhcNameOrFail name
-        thing <- GHC.lookupThing ghcName
-        pure (name, thing)
-    pure $ Map.fromList infos
-
--- | Strips all enclosing 'GHC.Tick's off an expression.
-stripTicks :: GHC.CoreExpr -> GHC.CoreExpr
-stripTicks = \case
-    GHC.Tick _ e -> stripTicks e
-    e -> e
 
 -- | Compiles all the marked expressions in the given binder into PLC literals.
-compileMarkedExprsBind :: (PluginOptions, GHC.FamInstEnvs) -> GHC.Name -> GHC.CoreBind -> GHC.CoreM GHC.CoreBind
-compileMarkedExprsBind opts markerName = \case
-    GHC.NonRec b e -> GHC.NonRec b <$> compileMarkedExprs opts markerName e
-    GHC.Rec bs -> GHC.Rec <$> mapM (\(b, e) -> (,) b <$> compileMarkedExprs opts markerName e) bs
-
--- | Compiles all the marked expressions in the given expression into PLC literals.
-compileMarkedExprs :: (PluginOptions, GHC.FamInstEnvs) -> GHC.Name -> GHC.CoreExpr -> GHC.CoreM GHC.CoreExpr
-compileMarkedExprs opts markerName =
-    let
-        comp = compileMarkedExprs opts markerName
-        compB = compileMarkedExprsBind opts markerName
-    in \case
+compileBind :: (?markerName :: GHC.Name, ?opts :: PluginOptions, ?famEnvs :: GHC.FamInstEnvs)
+            => GHC.CoreBind -> ECore PLC.DefaultUni GHC.CoreBind
+compileBind = \case
+    GHC.NonRec b rhs -> GHC.NonRec b <$> compileMarkers rhs
+    GHC.Rec bindsRhses -> GHC.Rec <$> for bindsRhses
+                         (\(b,rhs) -> (b,) <$> compileMarkers rhs)
+
+-- | Compiles all the core-expressions surrounded by plc in the given expression into PLC literals.
+compileMarkers :: (?markerName :: GHC.Name, ?opts :: PluginOptions, ?famEnvs :: GHC.FamInstEnvs)
```

Comment:
This is IMO a worse name than the previous one: we really do compile the *marked expressions*, not the markers!